subprojects/language-native/src/integTest/groovy/org/gradle/language/nativeplatform/ParallelNativePluginsIntegrationTest.groovy

```diff
@@ -140,5 +146,149 @@ class ParallelNativePluginsIntegrationTest extends AbstractInstalledToolChainInt
         installations.every { installation, app ->
             installation.exec().out == app.englishOutput
         }
+
+        and:
+        buildOperations.assertConcurrentOperationsExecuted(ExecuteTaskBuildOperationType)
+    }
+
+    def "can execute link executable tasks in parallel"() {
+        given:
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("link.*Executable")
+    }
+
+    def "can execute link shared library tasks in parallel"() {
+        given:
+        withComponentsForAppsAndSharedLibs(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("link.*SharedLibrary")
+    }
+
+    def "can execute create static library tasks in parallel"() {
+        given:
+        withComponentsForAppsAndStaticLibs(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("create.*StaticLibrary")
+    }
+
+    def "can execute compile tasks in parallel"() {
+        given:
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("compile.*")
+    }
+
+    def "can execute assemble tasks in parallel"() {
+        given:
+        apps = [
+            assemblerWithC1 : new AssemblerWithCHelloWorldApp(toolChain),
+            assemblerWithC2 : new AssemblerWithCHelloWorldApp(toolChain)
+        ]
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("assembleAssembler.*")
+    }
+
+    @RequiresInstalledToolChain(VISUALCPP)
+    def "can execute windows resource compile tasks in parallel"() {
+        given:
+        apps = [
+            windowsRC1 : new WindowsResourceHelloWorldApp(),
+            windowsRC2 : new WindowsResourceHelloWorldApp()
+        ]
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("compile.*Rc")
+    }
+
+    /**
+     * This is an approximation to check that a set of tasks execute in parallel by default.  Basically it checks to
+     * see if any of the tasks execute concurrently with any other task.  We can't guarantee that another task will
+     * always run concurrently while a given task is executing, so we settle for checking that the opposite does not
+     * happen (i.e. that none of the tasks execute concurrently with other tasks).
+     */
+    void assertTasksAreParallel(String regex) {
```

Comment:
i think this approach might lead to flakiness, as there's nothing that guarantees that the native tasks will in fact run concurrently with each other (only that it's likely that at least 2 will run at the same time). Instead, I would do something like add a dummy task to each project that blocks for the duration of the build, and assert that each native task runs concurrently with that, more like the Swift test case.